extendr-api/src/wrapper/s4.rs

```diff
@@ -0,0 +1,153 @@
+//! S4 class support.
+
+use super::*;
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct S4 {
+    pub(crate) robj: Robj,
+}
+
+impl S4 {
+    /// Make a new S4 type.
+    ///
+    /// Example:
+    /// ```
+    /// use extendr_api::prelude::*;
+    ///
+    /// test! {
+    ///     let robj = S4::new();
+    ///     assert_eq!(robj.rtype(), RType::S4);
+    /// }
+    /// ```
+    pub fn new() -> Self {
+        let robj = single_threaded(|| unsafe { new_owned(Rf_allocS4Object()) });
+        Self { robj }
+    }
```

Comment:
Is it legal to use an S4 object without specifying the class? Sorry, I almost know nothing about S4, especially how to handle it via C API.

Rcpp does require the class.

https://github.com/RcppCore/Rcpp/blob/423a2ed79bfda9b4e5e06b5dc1c646b02d60eb42/inst/include/Rcpp/S4.h#L57-L62

Reply:
Good point.

Likewise, I'm not proposing using S4 classes, but we should at least be able to read/write their slots.

There is a lot of redundancy in the S4 data model. The S4 flag, the S4SXP type and the class
attribute may all need to be there together.

Reply:
> we should at least be able to read/write their slots.

I see, agreed. Then, how about removing `new()` method (or at least `pub`)? Tests can use an S4 object created by `R!()` macro.